radix-engine/src/blueprints/epoch_manager/epoch_manager.rs

```diff
@@ -383,58 +399,178 @@ impl EpochManagerBlueprint {
+            .into_iter()
+            .map(|entry| (entry.component_address, entry.validator))
             .collect();
-        api.field_lock_write_typed(statistic_handle, statistic)?;
-        api.field_lock_release(statistic_handle)?;
 
+        // emit epoch change event
         Runtime::emit_event(
             api,
             EpochChangeEvent {
                 epoch,
-                validators: next_validator_set,
+                validators: next_validator_set.clone(),
             },
         )?;
 
+        // write zeroed statistics of next validators
+        statistic_substate.validator_statistics = (0..next_validator_set.len())
+            .map(|_index| ProposalStatistic::default())
+            .collect();
+        api.field_lock_write_typed(statistic_handle, statistic_substate)?;
+        api.field_lock_release(statistic_handle)?;
+
+        // write next validator set
+        validator_set_substate.validator_set = next_validator_set;
+        api.field_lock_write_typed(validator_set_handle, validator_set_substate)?;
+        api.field_lock_release(validator_set_handle)?;
+
         Ok(())
     }
+
+    /// Emits a configured XRD amount ([`EpochManagerConfigSubstate.total_emission_xrd_per_epoch`])
+    /// and distributes it across the given validator set, according to their stake.
+    fn emit_validator_rewards<Y>(
+        validator_set: BTreeMap<ComponentAddress, Validator>,
+        validator_statistics: Vec<ProposalStatistic>,
+        config: &EpochManagerConfigSubstate,
+        api: &mut Y,
+    ) -> Result<(), RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        let validator_rewards = validator_set
+            .into_iter()
+            .zip(validator_statistics)
+            .filter_map(|((address, validator), statistic)| {
+                ValidatorReward::create_if_applicable(
+                    address,
+                    validator.stake,
+                    statistic.success_ratio(),
+                    config.min_validator_reliability,
+                )
+            })
+            .collect::<Vec<_>>();
+
+        if validator_rewards.is_empty() {
+            return Ok(());
+        }
+
+        let stake_sum_xrd = validator_rewards
+            .iter()
+            .map(|validator_reward| validator_reward.stake_xrd)
+            .sum::<Decimal>();
+        let emission_per_staked_xrd = config.total_emission_xrd_per_epoch / stake_sum_xrd;
+        let effective_total_emission_xrd = validator_rewards
+            .iter()
+            .map(|validator_reward| validator_reward.effective_stake_xrd * emission_per_staked_xrd)
+            .sum::<Decimal>();
+
+        let total_emission_xrd_bucket =
+            ResourceManager(RADIX_TOKEN).mint_fungible(effective_total_emission_xrd, api)?;
+
+        for validator_reward in validator_rewards {
+            let emission_xrd_bucket = total_emission_xrd_bucket.sys_take(
```

Comment:
Maybe better would be to just mint per validator inside the for loop?

The `sys_drop_empty` feels fragile. Even though theoretically speaking using fixed point decimal and correct math means this should always succeed, would be nicer to not rely on this invariant.

Reply:
I actually wanted to have a single `Mint` event because:
- it is easy to interpret: "here an emission happened on epoch..."
  - a `RewardReceivedEvent` is going to be another per-validator event (see [Q10 here](https://radixdlt.atlassian.net/wiki/spaces/S/pages/3058827265/REP-83+Calculating+Emissions+and+Handling+Validator+Fees)) and it very nicely continues the interpretation with "... and this share of it was given to that validator".
- it doesn't inflate the "epoch changed" on ledger

I definitely would think twice if calculating the validator's share was subject to rounding errors. 
(then it would be really risky to assert a non-empty bucket 😨 )

But here I only rely on `a + b + c + ... = S` <=> `S - a - b - c - ... = 0` which seems safe enough on a fixed-point math.

Please let me know if you prefer to have some mitigations here:
A) we could "just give the entire remaining bucket to the last validator on the list"; (seems "unfair")
B) or we could "burn any non-0 XRD left in that bucket"; (seems funny; also: what if there is too little XRD in the bucket?)
C) or maybe do `S - c - b - a - ... = 0` instead? (somehow it feels more "stable"; I mean, maybe even the regular floating-point math is reversible this way?)

_(anyways, we can adjust it in some next PR)_